apache/trafficcontrol
# Converts Edge/Mid header rewrite, regexremap, remaptext and tr request/response header text inputs to textareas

- [x] This PR fixes #3426 

By converting the edge/mid header rewrite, regex remap, remap text and tr request/response header fields of the DS forms to textareas instead of text input, line breaks can be preserved and eliminate the need to use an explicit `__RETURN__` to indicate a line break. (although `__RETURN__` will still be respected and recognized as a line break for backwards compatability).

This PR also changes the presentation of delivery service request diffs to account for multiline values.

## Which Traffic Control components are affected by this PR?

- Traffic Portal

## What is the best way to verify this PR?

1. Run automated tests

Run the TO unit tests -

cd trafficcontrol/traffic_ops/traffic_ops_golang
go test $(go list ./...)

Run the API tests - https://github.com/apache/trafficcontrol/blob/master/traffic_ops/testing/api/README.md

Run the UI tests - https://github.com/apache/trafficcontrol/tree/master/traffic_portal/test/end_to_end

2. Manual validation

a. set dsRequests.enabled=true in traffic_portal_properties.json and restart TP
b. in TP, create/edit a dns* (you can do http*, steering or any_map as well but only some field apply to each) delivery service and modify the edge/mid header rewrite rules, regex remap, remap text and tr request/response headers to include multiline rules.
c. save the changes which will create a ds request.
d. review the ds request and fulfill it. fulfilling the ds request persists the ds changes to the database.
e. verify the contents of the header rewrite files and ensure the multiline rules are presented on separate lines:

edge header rewrite rules can be viewed by GET api/1.4/cdns/cdn-name/configfiles/ats/hdr_rw_xml-id.config
mid header rewrite rules can be view by GET api/1.4/cdns/cdn-name/configfiles/ats/hdr_rw_mid_xml-id.config

f. verify the contents of the regex remap config file and ensure the multiline rules are presented on separate lines:

GET api/1.4/cdns/cdn-name/configfiles/ats/regex_remap_xml-id.config

g. verify the tr request/response rules show up in the snapshot correctly

- add multiline rules for tr request/response headers
- do a snapshot diff in TP, look in the ds section for your tr request/response header change. ensure that multiple lines resulted in multiple header rules
- perform a snapshot then you can inspect manually if you want as follows:

GET api/1.4/cdns/cdn-name/snapshot (current snapshot)
GET api/1.4/cdns/cdn-name/snapshot/new (next snapshot)

Note: we don't currently have any API tests to parse and validate the contents of config files.

## The following criteria are ALL met by this PR
